biopet-core/src/main/scala/nl/lumc/sasc/biopet/core/report/ReportBuilder.scala

```diff
@@ -223,62 +247,110 @@ trait ReportBuilder extends ToolCommand {
-      case (name, subPage) => Some(generatePage(summary, subPage, outputDir, path ::: name :: Nil, pageArgs))
-      case _               => None
+    def pageArgs(page: ReportPage) = {
+      val rootPath = "./" + Array.fill(path.size)("../").mkString
+      args ++ page.args ++
+        Map("page" -> page,
+          "path" -> path,
+          "outputDir" -> pageOutputDir,
+          "rootPath" -> rootPath,
+          "allPipelines" -> pipelines,
+          "allModules" -> modules,
+          "allSamples" -> samples,
+          "allLibraries" -> libraries
+        )
     }
 
-    val output = ReportBuilder.renderTemplate("/nl/lumc/sasc/biopet/core/report/main.ssp",
-      pageArgs ++ Map("args" -> pageArgs))
+    val subPageJobs = pageFuture.map { page =>
+      // Generating subpages
+      page.subPages.flatMap {
+        case (name, subPage) => generatePage(summary, subPage, outputDir, path ::: name :: Nil, pageArgs(page))
+      }
+    }
 
-    val file = new File(pageOutputDir, "index.html")
-    val writer = new PrintWriter(file)
-    writer.println(output)
-    writer.close()
+    val renderFuture = pageFuture.map { page =>
+      pageOutputDir.mkdirs()
 
-    done += 1
-    if (done % 100 == 0) logger.info(done + " Done, " + (done.toDouble / total * 100) + "%")
-    jobs.sum + 1
+      val file = new File(pageOutputDir, "index.html")
+      logger.info(s"Start rendering: $file")
+
+      val output = ReportBuilder.renderTemplate("/nl/lumc/sasc/biopet/core/report/main.ssp",
+        pageArgs(page) ++ Map("args" -> pageArgs(page)))
+
+      val writer = new PrintWriter(file)
+      writer.println(output)
+      writer.close()
+      logger.info(s"Done rendering: $file")
+
+    }
+
+    renderFuture :: Await.result(subPageJobs, Duration.Inf)
   }
+
+  def pipelineName: String
+
+  /** Files page, can be used general or at sample level */
+  def filesPage(sampleId: Option[Int] = None, libraryId: Option[Int] = None): Future[ReportPage] = {
+    val dbFiles = summary.getFiles(runId, sample = sampleId.map(SampleId),
+      library = libraryId.map(LibraryId))
+      .map(_.groupBy(_.pipelineId))
+    val modulePages = dbFiles.map(_.map {
+      case (pipelineId, files) =>
+        val moduleSections = files.groupBy(_.moduleId).map {
+          case (moduleId, files) =>
+            val moduleName: Future[String] = moduleId match {
+              case Some(id) => summary.getModuleName(pipelineId, id).map(_.getOrElse("Pipeline"))
+              case _        => Future("Pipeline")
+            }
+            moduleName.map(_ -> ReportSection("/nl/lumc/sasc/biopet/core/report/files.ssp", Map("files" -> files)))
+        }
+        val moduleSectionsSorted = moduleSections.find(_._1 == "Pipeline") ++ moduleSections.filter(_._1 != "Pipeline")
+        summary.getPipelineName(pipelineId = pipelineId).map(_.get -> Future(ReportPage(Nil, Await.result(Future.sequence(moduleSectionsSorted), Duration.Inf).toList, Map())))
```

Comment:
Instead of `Future(ReportPage(...` this should also be `Future.successful(ReportPage..`. More importantly, this relates to the previous question on `ReportPage`: if we make its `subPage` argument just a `List[(String, ReportPage)]`, we can avoid doing the `Await` call here.